packages/http/src/validator/index.ts

```diff
@@ -24,65 +27,82 @@ export const queryValidator = new HttpQueryValidator(queryDeserializerRegistry,
 export const pathValidator = new HttpPathValidator(pathDeserializerRegistry, 'path');
 
 const validateInput: ValidatorFn<IHttpOperation, IHttpRequest> = ({ resource, element }) => {
-  const results: IPrismDiagnostic[] = [];
   const mediaType = caseless(element.headers || {}).get('content-type');
-
-  // Replace resource.request in this function with request
   const { request } = resource;
-
   const { body } = element;
-  if (request && request.body) {
-    if (!body && request.body.required) {
-      results.push({ code: 'required', message: 'Body parameter is required', severity: DiagnosticSeverity.Error });
-    } else if (body) {
-      bodyValidator
-        .validate(body, (request && request.body && request.body.contents) || [], mediaType)
-        .forEach(validationResult => results.push(validationResult));
-    }
-  }
 
-  return results
-    .concat(headersValidator.validate(element.headers || {}, (request && request.headers) || []))
-    .concat(queryValidator.validate(element.url.query || {}, (request && request.query) || []))
-    .concat(pathValidator.validate(getPathParams(element.url.path, resource.path), (request && request.path) || []));
+  const validateBody = pipe(
+    Option.fromNullable(request),
+    Option.mapNullable(request => request.body),
+    Option.chain(requestBody =>
+      pipe(
+        requestBody,
+        Option.fromPredicate(requestBody => !!requestBody.required && !body),
+        Option.map<unknown, NonEmptyArray<IPrismDiagnostic>>(() => [
+          { code: 'required', message: 'Body parameter is required', severity: DiagnosticSeverity.Error },
+        ]),
+        Option.alt(() =>
+          pipe(
+            sequenceOption(Option.fromNullable(body), Option.fromNullable(requestBody.contents)),
+            Option.chain(([body, contents]) =>
+              pipe(
+                bodyValidator.validate(body, contents, mediaType),
+                Either.swap,
+                Option.fromEither
+              )
+            )
+          )
+        )
+      )
+    ),
+    Either.fromOption(() => {}),
+    Either.swap
+  );
+
+  return pipe(
+    sequenceValidation(
+      validateBody,
```

Comment:
would it be feasible for the composition of `validateBody` to be put outside of `validateInput`?

Reply:
We can definitely do that but why?

Reply:
Composing `validateBody` and running the sequence seems like two responsibilities to me. If we can spot a distinct piece of code, that we could move out and hence lower the complexity of `validateInput`, then why not go for it?

Reply:
Aaaaaaah now I get what you mean. It totally makes sense! Will do

Reply:
Done!